File: src/leetcode/java/medium/UTF8Validation_393.java
package leetcode.java.medium;

public class UTF8Validation_393 {
  public boolean validUtf8(int[] data) {
    int count = 0;
    for (int d : data) {
      if (count == 0) {
        if ((d >> 5) == 0b110) count = 1;
        else if ((d >> 4) == 0b1110) count = 2;
        else if ((d >> 3) == 0b11110) count = 3;
        else if ((d >> 7) == 0b1) return false;
      } else {
        if ((d >> 6) != 0b10) return false;
        count--;
      }
    }
    return count == 0;
  }

  public static void main(String[] args) {
    UTF8Validation_393 ust = new UTF8Validation_393();
    int[] data = { 255 };
    System.out.println(ust.validUtf8(data));
  }
}

File: src/leetcode/python/UTF8Validation_393.py
class UTF8Validation_393(object):
    def validUtf8(self, data):
        """
        :type data: List[int]
        :rtype: bool
        """
        c = 0
        for d in data:
            if c == 0:
                if d >> 5 == 0b110:
                    c = 1
                elif d >> 4 == 0b1110:
                    c = 2
                elif d >> 3 == 0b11110:
                    c = 3
                elif d >> 7 == 0b1:
                    return False
            else:
                if d >> 6 != 0b10: return False
                c -= 1
        return c == 0
